crankshaft-engine/src/service/runner/backend/docker.rs

```diff
@@ -35,11 +97,156 @@ impl Backend {
+                            .unwrap()
+                            .resources
+                            .as_ref()
+                            .unwrap();
+
+                        let node_cpu: u64 = resources
+                            .nano_cpus
+                            .map(|n| n / 1_000_000_000)
+                            .unwrap_or(0)
+                            .try_into()
+                            .context("node CPU count is negative")?;
+                        swarm.cpu += node_cpu;
+                        swarm.max_cpu = swarm.max_cpu.max(node_cpu);
+
+                        let node_memory: u64 = resources
+                            .memory_bytes
+                            .unwrap_or(0)
+                            .try_into()
+                            .context("node memory is negative")?;
+                        swarm.memory += node_memory;
+                        swarm.max_memory = swarm.max_memory.max(node_memory);
+
+                        debug!(
+                            id = node
+                                .id
+                                .as_ref()
+                                .context("Docker daemon reported a node without an identifier")?,
+                            total_cpu = node_cpu,
+                            total_memory = node_memory,
+                            "found Docker swarm node"
+                        );
+                    }
+
+                    if swarm.nodes == 0 {
+                        bail!(
+                            "the local Docker daemon is part of a swarm but there are no active \
+                             and ready nodes"
+                        );
+                    }
+
+                    Some(swarm)
+                }
+                (Some(id), _) if !id.is_empty() => {
+                    bail!(
+                        "the local Docker daemon is part of a swarm but the node state is not \
+                         active"
+                    );
+                }
+                _ => {
+                    // Not part of a swarm
+                    None
+                }
+            }
+        } else {
+            None
+        };
+
+        let resources = match swarm {
+            Some(swarm) => {
+                info!(
+                    nodes = swarm.nodes,
+                    cpu = swarm.cpu,
+                    memory = swarm.memory,
+                    max_cpu = swarm.max_cpu,
+                    max_memory = swarm.max_memory,
+                    "Docker backend is interacting with a swarm"
+                );
+
+                Resources::Swarm(swarm)
+            }
+            None => {
+                let cpu = info
+                    .ncpu
+                    .map(|n| {
+                        n.try_into()
+                            .context("Docker daemon reported a negative CPU count")
+                    })
+                    .transpose()?
+                    .unwrap_or(0);
```

Comment:
I wonder if this should be an `expect()`? It seems like we'd want the user to know and to handle that case rather than defaulting to essentially not using the node's capacity (or, at least, that's what I'm interpreting this more or less does).